{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d6c66aa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import statistics\n",
    "from sklearn import preprocessing\n",
    "import math\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.ensemble import ExtraTreesClassifier\n",
    "\n",
    "from numpy import mean\n",
    "from numpy import std\n",
    "from sklearn.metrics import  confusion_matrix\n",
    "from imblearn.over_sampling import SMOTE\n",
    "from sklearn.svm import LinearSVC\n",
    "from sklearn.feature_selection import SelectFromModel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "bdb7cd19",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "#Function to transpose genomic matrice, df.T can also be employed but computionally heavy for my personal laptop\n",
    "def transpose_ (df_gene_full__FINAL_):\n",
    "    id_gene = df_gene_full__FINAL_.columns\n",
    "    id_gene_list = list(id_gene)\n",
    "\n",
    "    Transpose_Gene = []\n",
    "    for i_transpose in id_gene_list :\n",
    "        b_column = df_gene_full__FINAL_[i_transpose]\n",
    "        b_row = np.array(b_column)\n",
    "        Transpose_Gene.append(b_row)\n",
    "    return Transpose_Gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e7bab017",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#function to detele last letter of patients name in genomic set to compare both set\n",
    "def new_colum ():\n",
    "    list_without_A = []\n",
    "    df_gene_1_sorted =df_gene_1.df[sorted(df_gene_1.df.columns)]\n",
    "    for name_TCGA in (list(df_gene_1_sorted.columns)):\n",
    "        newString = name_TCGA.rstrip(name_TCGA[-1])\n",
    "        list_without_A.append(newString)\n",
    "    df_gene_1_sorted.set_axis(list_without_A, axis=1)\n",
    "    \n",
    "    return df_gene_1_sorted,list_without_A\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "678efc68",
   "metadata": {},
   "outputs": [],
   "source": [
    "#function to delete in the clinical set, patients not present in genomic set.\n",
    "# Index to drop genes, were typed manually, from an issue with the type of variable, but can be retrieved with line index_genes.\n",
    "def id_iden(list_without_A):\n",
    "    list_name_not_present = []\n",
    "    Id_TCGA__ =df_clin_['Sample ID']\n",
    "    Id_TCGA_list__ =list(Id_TCGA__)\n",
    "    list_a = []\n",
    "    for indice_TCGA in range(0,len(Id_TCGA_list__)) :\n",
    "        if Id_TCGA_list__[indice_TCGA] not in list_without_A :\n",
    "            list_a.append(Id_TCGA_list__[indice_TCGA])\n",
    "            \n",
    "          #  index_genes = df_clin_.index[Id_TCGA__ == Id_TCGA_list__[indice_TCGA]].astype(int)\n",
    "            v = df_clin_.iloc[indice_TCGA].name\n",
    "            b = v.tolist()\n",
    "    df_clin__ =  df_clin_.drop(222)\n",
    "   \n",
    "    df_clin__ =  df_clin__.drop(537)\n",
    "    df_clin__ =  df_clin__.drop(1009)\n",
    "    return df_clin__\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "88e456c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to analyse both set and compare their siilarity, so patients are in the same order\n",
    "\n",
    "\n",
    "def size_simi (df_gene_1_,df_clin__):\n",
    "    \n",
    "    Id_TCGA =df_clin__['Sample ID'] \n",
    "    Id_TCGA_list =list(Id_TCGA)\n",
    "  \n",
    "    id_gene = df_gene_1_.columns\n",
    "    id_gene_list = list(id_gene)\n",
    "\n",
    "    df_gene_1_sorted_full =df_gene_1.df[sorted(df_gene_1.df.columns)]\n",
    "    id_gene_full = df_gene_1_sorted_full.columns\n",
    "    id_gene_full_list = list(id_gene_full)\n",
    "\n",
    "    \n",
    "\n",
    "    list_not_present_gene = []\n",
    "    list_not_present_gene_full = []\n",
    "    list__present_gene = []\n",
    "    \n",
    "    for id_gene in range (0, len(id_gene_list)):\n",
    "        if id_gene_list[id_gene] not in Id_TCGA_list:\n",
    "            \n",
    "            list_not_present_gene.append(id_gene_list[id_gene])\n",
    "            list_not_present_gene_full.append(id_gene_full_list[id_gene])         \n",
    "       \n",
    "    df_gene_1_ =  df_gene_1_.drop(columns= list_not_present_gene )\n",
    "    df_gene_full =  df_gene_1_sorted_full.drop(columns= list_not_present_gene_full )\n",
    "\n",
    "    list_double = []\n",
    "    id_gene_1 = df_gene_1_.columns\n",
    "    id_gene_list_1 = list(id_gene_1)\n",
    "    id_gene_1_full= df_gene_full.columns\n",
    "    \n",
    "    for id_TCGA in range (0, len(Id_TCGA_list )):    \n",
    "        id_drop_gene = id_TCGA\n",
    "        while Id_TCGA_list[id_TCGA]!=id_gene_list_1[id_drop_gene] :\n",
    "            name_column = df_gene_full.columns[id_drop_gene]\n",
    "            df_gene_full =df_gene_full.drop(columns = name_column)\n",
    "            id_gene_list_1.remove(id_gene_list_1[id_drop_gene])\n",
    "          \n",
    "\n",
    "    return df_clin__,df_gene_full"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "a9154e64",
   "metadata": {},
   "outputs": [],
   "source": [
    "#First attribute selection, by deleting genes with variance inbetween patients below 2, as non predictive for classification\n",
    "def  variance_clean(df_gene_final__):\n",
    "\n",
    "    list_var = []\n",
    "    df_gene_final__ = df_gene_final__.replace(to_replace=0,\n",
    "           value=0.8)\n",
    "    df_gene_final__ = np.log2(df_gene_final__)\n",
    "    df_gene_final__var = df_gene_final__.var(axis=1)\n",
    "    a = df_gene_final__var[df_gene_final__var < 2.1].dropna()\n",
    "    df_gene_final__ =  df_gene_final__.drop( labels =a.index)\n",
    "  \n",
    "    return df_gene_final__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "be50b8a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Read_TSV:\n",
    "     def __init__(self, name_file):\n",
    "        self.name_file = name_file\n",
    "        \n",
    "     def get_tsv(self):\n",
    "        self.df= pd.read_csv(self.name_file, sep='\\t')\n",
    "      \n",
    "\n",
    "     def get_shape(self):\n",
    "        self.row_gene, self.col_gene = self.df.shape\n",
    "       \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "c5ab14fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Genomic set lecture\n",
    "df_gene_1 = Read_TSV('TCGA_BRCA_tpm.tsv')\n",
    "df_gene_1.get_tsv()\n",
    "df_gene_1.get_shape()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "0b472946",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Clinical set elcture\n",
    "df_clin = Read_TSV('brca_tcga_pan_can_atlas_2018_clinical_data_filtered.tsv')\n",
    "df_clin.get_tsv()\n",
    "df_clin.get_shape()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "b0d017fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Delete patients with unknown subtypes\n",
    "nan_rows = df_clin.df.isnull().any(axis=1)\n",
    "labels_nan_rows = df_clin.df[nan_rows].index\n",
    "\n",
    "df_clin_ =  df_clin.df.drop( labels =labels_nan_rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "31f1f352",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\users\\laura\\appdata\\local\\programs\\python\\python36\\lib\\site-packages\\ipykernel_launcher.py:8: FutureWarning: set_axis currently defaults to operating inplace.\n",
      "This will change in a future version of pandas, use inplace=True to avoid this warning.\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "#\n",
    "df_gene_final,list_drop_A = new_colum ( )\n",
    "df_clin_final = id_iden(list_drop_A)\n",
    "df_clin__FINAL,df_gene_full__2 = size_simi(df_gene_final,df_clin_final)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "8a70fe10",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Feature selection to eliminate non predictive gene\n",
    "df_gene_full__FINAL = variance_clean(df_gene_full__2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "65dcb6cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Matrix transposition\n",
    "df_gene_full__FINAL_transpose = transpose_ (df_gene_full__FINAL)\n",
    "df_gene_full__FINAL_transpose = pd.DataFrame(df_gene_full__FINAL_transpose)\n",
    "\n",
    "#Label column selection\n",
    "label = df_clin__FINAL['Subtype']\n",
    "a = np.array(label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "1f176ae8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAEzCAYAAADKCUOEAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAYH0lEQVR4nO3de7RmdX3f8ffHGRSjCCgDiwWjYyKpYrxARpwEewnYBJQIbcGKVtBiJ6sxq7bmUmLNRU0qZjVq8bbWRFS0VsVUKytaogs0bWJRB0GEoGWKIFOQGQXxQrwA3/7x7OO5zHPmnJk559nP+e33a61ZZ+/f3jPnO78589n7+e3f3jtVhSSpLQ/puwBJ0soz3CWpQYa7JDXIcJekBhnuktQgw12SGrS+7wIAjjjiiNq0aVPfZUjSmnLNNdd8s6o2jNs2FeG+adMmtm/f3ncZkrSmJLltsW0Oy0hSgwx3SWqQ4S5JDVpWuCe5NcmXk1yXZHvX9ugkn0pyc/f18K49SS5OsiPJ9UlOXM2/gCRpT/ty5v5LVfX0qtrcrV8IXFlVxwFXdusApwPHdb+2Au9YqWIlSctzIMMyZwKXdsuXAmfNaX9vjVwNHJbk6AP4PpKkfbTccC/gk0muSbK1azuqqu4E6L4e2bUfA9w+5/fu7NokSROy3HnuJ1fVHUmOBD6V5Ct72Tdj2vZ4aHx3kNgK8NjHPnaZZUiSlmNZ4V5Vd3RfdyX5KHAScFeSo6vqzm7YZVe3+05g45zffixwx5g/cxuwDWDz5s0H/MaQTRd+/ED/iAN260XP7bsESQKWMSyT5BFJDplZBn4ZuAG4HDi/2+184GPd8uXAed2smS3AvTPDN5KkyVjOmftRwEeTzOz/X6vqiiRfAC5LcgHwdeCcbv9PAM8BdgD3AS9d8aolSXu1ZLhX1S3A08a0fws4dUx7AS9fkeokSfvFO1QlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalByw73JOuSXJvkL7r1xyf5XJKbk3woyUO79od16zu67ZtWp3RJ0mL25cz9FcBNc9bfALypqo4D7gEu6NovAO6pqicAb+r2kyRN0LLCPcmxwHOBd3brAU4B/rzb5VLgrG75zG6dbvup3f6SpAlZ7pn7m4HfAR7s1h8DfLuq7u/WdwLHdMvHALcDdNvv7fafJ8nWJNuTbN+9e/d+li9JGmfJcE9yBrCrqq6Z2zxm11rGttmGqm1VtbmqNm/YsGFZxUqSlmf9MvY5GXhekucABwOPYnQmf1iS9d3Z+bHAHd3+O4GNwM4k64FDgbtXvHJJ0qKWPHOvqt+tqmOrahPwAuCqqnoR8Gng7G6384GPdcuXd+t026+qqj3O3CVJq+dA5rn/e+CVSXYwGlO/pGu/BHhM1/5K4MIDK1GStK+WMyzzE1X1GeAz3fItwElj9vkBcM4K1CZJ2k/eoSpJDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ1aMtyTHJzk80m+lOTGJK/p2h+f5HNJbk7yoSQP7dof1q3v6LZvWt2/giRpoeWcuf8QOKWqngY8HTgtyRbgDcCbquo44B7ggm7/C4B7quoJwJu6/SRJE7RkuNfI97rVg7pfBZwC/HnXfilwVrd8ZrdOt/3UJFmxiiVJS1rWmHuSdUmuA3YBnwL+L/Dtqrq/22UncEy3fAxwO0C3/V7gMWP+zK1JtifZvnv37gP7W0iS5llWuFfVA1X1dOBY4CTgSeN2676OO0uvPRqqtlXV5qravGHDhuXWK0lahn2aLVNV3wY+A2wBDkuyvtt0LHBHt7wT2AjQbT8UuHslipUkLc9yZstsSHJYt/xw4NnATcCngbO73c4HPtYtX96t022/qqr2OHOXJK2e9UvvwtHApUnWMToYXFZVf5Hkb4EPJvkj4Frgkm7/S4D3JdnB6Iz9BatQtyRpL5YM96q6HjhhTPstjMbfF7b/ADhnRaqTJO0X71CVpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNWjJcE+yMcmnk9yU5MYkr+jaH53kU0lu7r4e3rUnycVJdiS5PsmJq/2XkCTNt5wz9/uB36yqJwFbgJcnOR64ELiyqo4DruzWAU4Hjut+bQXeseJVS5L2aslwr6o7q+qL3fJ3gZuAY4AzgUu73S4FzuqWzwTeWyNXA4clOXrFK5ckLWqfxtyTbAJOAD4HHFVVd8LoAAAc2e12DHD7nN+2s2uTJE3IssM9ySOB/wb826r6zt52HdNWY/68rUm2J9m+e/fu5ZYhSVqG9cvZKclBjIL9/VX1ka75riRHV9Wd3bDLrq59J7Bxzm8/Frhj4Z9ZVduAbQCbN2/eI/y1/zZd+PG+S+DWi57bdwnSoC1ntkyAS4CbquqNczZdDpzfLZ8PfGxO+3ndrJktwL0zwzeSpMlYzpn7ycCLgS8nua5rexVwEXBZkguArwPndNs+ATwH2AHcB7x0RSuWJC1pyXCvqr9m/Dg6wKlj9i/g5QdYlyTpAHiHqiQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDVrW4wckqSVDeESHZ+6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1KAlwz3Ju5LsSnLDnLZHJ/lUkpu7r4d37UlycZIdSa5PcuJqFi9JGm85Z+7vAU5b0HYhcGVVHQdc2a0DnA4c1/3aCrxjZcqUJO2LJcO9qv4ncPeC5jOBS7vlS4Gz5rS/t0auBg5LcvRKFStJWp79HXM/qqruBOi+Htm1HwPcPme/nV3bHpJsTbI9yfbdu3fvZxmSpHFW+oJqxrTVuB2raltVba6qzRs2bFjhMiRp2PY33O+aGW7pvu7q2ncCG+fsdyxwx/6XJ0naH/sb7pcD53fL5wMfm9N+XjdrZgtw78zwjSRpctYvtUOSDwD/CDgiyU7gD4CLgMuSXAB8HTin2/0TwHOAHcB9wEtXoWZJ0hKWDPeqOneRTaeO2beAlx9oUZKkA+MdqpLUIMNdkhpkuEtSgwx3SWrQkhdUJbVh04Uf77sEbr3ouX2XMBieuUtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGuTLOtQ0X1ChofLMXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJatCqhHuS05J8NcmOJBeuxveQJC1uxcM9yTrgbcDpwPHAuUmOX+nvI0la3GqcuZ8E7KiqW6rqR8AHgTNX4ftIkhaRqlrZPzA5Gzitql7Wrb8YeGZV/caC/bYCW7vVvwd8dUUL2T9HAN/su4gpYV+M2A+z7ItZ09IXj6uqDeM2rMbz3DOmbY8jSFVtA7atwvffb0m2V9XmvuuYBvbFiP0wy76YtRb6YjWGZXYCG+esHwvcsQrfR5K0iNUI9y8AxyV5fJKHAi8ALl+F7yNJWsSKD8tU1f1JfgP4S2Ad8K6qunGlv88qmaphop7ZFyP2wyz7YtbU98WKX1CVJPXPO1QlqUGGuyQ1yHCXpAYZ7pLUoNW4iWlNS3Iy8MKqennftfQpyRHAt2pgV9yTfJkxN90xujmvquqpEy6pN91zol7G6F6VK6rqb+Zse3VV/VFvxU3YWvy5cLYMkOTpwAuB5wNfAz5SVW/pt6rJSbIFuAi4G3gd8D5Gt1c/BDivqq7osbyJSvK4vW2vqtsmVUvfkrwT+Cng88CLgb+qqld2275YVSf2Wd8krcWfi8GGe5KfZXSD1bnAt4APAb9VVXv9R2xRku3Aq4BDGc3fPb2qrk7yROADVXVCrwWqF0munzkjTbIeeDujg/65wNX+XEy3IY+5fwU4FfjVqnpWd6b+QM819WV9VX2yqj4MfKOqrgaoqq/0XFdvkmxJ8oUk30vyoyQPJPlO33VN2ENnFqrq/qraClwHXAU8sreqerSWfi6GHO7/DPgG8Okkf5bkVMY/9GwIHpyz/HcLtg3zox28ldEZ6s3AwxmNPQ9mqK6zPclpcxuq6rXAu4FNvVTUvzXzczHYYZkZSR4BnMXoH+wU4FLgo1X1yV4Lm6AkDwDfZ3Rwezhw38wm4OCqOqiv2voy89S/BUMTn62qX+y7tklK8hBgS1V9tu9apsFa+rkY/GyZqvo+8H7g/UkeDZwNXAgMJtyral3fNUyh+7oH312X5E+AO4FH9FzTxFXVg0n+FPiFvmuZEmvm52LwZ+4ASQ5n9JjinxzsquqL/VXUH/tipJsdsQs4CPh3jC42v72qdvRaWA+SvAa4ntEsskEHRvdzcRej6xFT/XMx+HBP8jrgJcAtzI49V1Wd0ltRPbEvNE6S7zI6O32A0TWZmbndj+q1sJ7NnAhV1fV91zKO4Z58FXhK977XQbMvZiU5g9Gc/8cx+hRjoIkknwGex+hn4jpgN3Pm/0+TIc+WmXEDcFjfRUwJ+2LWm4HzgcdU1aOq6pChBntG/kWS3+vWNyY5qe+6enJoVX0H+KfAu6vq54Fn91zTWIO/oAq8Hrg2yQ3AD2caq+p5/ZXUG/ti1u3ADUMfY+68ndEw3SmMPs18D3gb8Iw+i+rJ+iRHM7qb/T/0XczeGO6jqY9vAL7M/PneQ2RfzPod4BNJ/or5B7o39ldSb55ZVScmuRagqu7pZowM0WsZvWXur6vqC0l+mtGc96ljuMM3q+rivouYEvbFrD9mdIZ6MHPu1ByoH3cPESuAJBsY6MG/u4v7w3PWb2F0Q+TU8YJq8kZGZ2aXM/8MbYjT/+yLzszNKn3XMQ2SvAj458CJjD7dnQ28ugu6QUlyMHAB8GRGB34Aqupf9lbUIgz35NNjmgc5/c++mJXkIuCqId2pvDfdQ+RmHtFxZVXd1HNJvUjyYUbPpXohoyGaFwE3VdUrei1sjMGHuzTOnLndPwR+zACnQnZ3bC+qqu6eVC3TIsm1VXXCzOMHkhwE/OU0ngANfsw9ye+Pa+8ekDQo9sVIkgBPrqqv911Lz65hNM4e4Gjgjq49XftP91RXn37cff12kp9j9PDBTf2Vs7jBhzujB2bNOBg4AxjkR07sC2B0ep7ko8DP911Ln6rq8TPLM2esfdYzJbZ1d6b+HqNrU4/slqeOwzILJHkYcHlV/UrftfRtyH2R5G3Ae6rqC33XMg2G9ualFnjmvqefYpgfN8cZcl/8EvBrSW5j9nHIU/muTE1Gkn8I3FNV1yd5PvAPgB3AO6rqh3v/3ZM3+HBf8OLbdcAGRnfhDY59Mc/pfRfQtyRzn5dy5IL1Qd3Q1X2SeypwcPcMpkcCVwC/CLyL0ayZqTL4YZkFL769H7irqu7vq54+2RfzJXka8Pe71f9VVV/qs55JS/IHe9teVa+ZVC19S/K3VXV8N8/9/wFHVtUD3cX366vqKT2XuIfBn7mPe2t5kq9X1WP7qKdP9sWsJK8A/hXwka7pvyTZ1r1rdxCWG95JfreqXr/a9fTsBwBV9YMkt1XVA916Jfnx3n9rPwYf7osY6rtUxxlqX1zA6Jkq3wdI8gbgfzOl78vs2TmMHjrXsplhqTB/iCqMhi+njuE+3rDHquYbal+E0cspZjzAcA90SxlCv/wZcMiYZYB3Tr6cpQ023BdeHJq7idHFksGwL8Z6N/C5br47jF6ifkmP9Uyz5k8A1uIQ1WDDnflH3oX+88SqmA72xQJV9cbucb8nMzrIvbSqru25rGk1hDP35ZqaIarBz5ZZyjQdiftmX2hGkmfM3OCV5FVV9R/7rmkaTNOdvIb7Erwzb9YQ+qJ7YNjMf4rMWV4PPLSqBvtpN8nxwAuAc4F7fSTynqbp/8hgf1D3gR85ZzXfF1U1b4gqySHArwO/Bnx07G9qWHfvw7ndr/sZvTB8c1Xd2mddU2xq/o/4guyl+dFm1mD6IslhSf4Q+BKjaxLPqKrf7LeqyUryWeATwEHA2d3LoL9rsM+XZO67ZKfmBSaG+9Km5kg8BZrviyRHJHk98EVGZ6onVNWrq+pbPZfWh92MDmxHMTuXezAH+L1JcnyS1ya5GXjHTPs0XXtwWGaMuReLmKIjcR8G2Be3MQq1dwP3AReM7jAfGdLzVKrqzCSHMnpH6GuSPAE4LMlJVfX5nsubuLU2ROUF1Y4Xi2YNuS+6oZhF/1MM6XkqCyU5ktG7VM8FNlbVxp5LmphuiOpQ4IPAB6vq5iRfm/vM+2kz6DP3tXYkXk32xUhV/eFy9hvCtNDuIVmHVNVugKraBbwlyWXAXl/B16DdwLHMDlHdzJQPUQ12zN2LRbPsi/1yTt8FTMDFzD4Vc65nA/9mwrX0qqrOBJ7C6FrMa5J8DTg8yUn9Vra4wYY7Xiyay77Yd81fXAaeVVUfWdhYVe9n9KKKQamqe6vqXVX1j4FnAr8PvDnJ7T2XNtZgw30tHolXi32xX4Zw8NvbAWxQ2ZHk4CQ/efpjVe3qHv/8T4Bf7q+yxQ3qH2ihtXYkXk32xT4bwpn7rnEH+G5e9+4e6unTmhuiGuxsmYUXi+a0HwU8uqpu6qeyybMvlmdoz1Ppgv0y4D3ANV3zZuA84AVV9bmeSpu4mTcxLbLtxqp68qRrWsqQz9zX3JF4FdkXi1gLN6uslm4u+zMZfUp5SfcrjF5iMphg76y5Iaohn7mvuSPxarEv5nNa6NKSnFxVf9N3HZPSPf75txfevNUNUf1pVU3dBeYhz3Nfc0fiVWRfdBbcrHL2nJtVbu23sslLsg54PnAM8D+q6sYkZwCvAh4OTMWjbSfkt4HLkryHMUNUfRW1N4P6j7uAF4tm2ReznBY66xLgZcBjGN289G7gPwF/Mi3PLJ+UtThENeRhGS8WdeyL+eY8T+Vc4AnAYcCvDO15KkluAJ5aVQ92F92/CTyhqr7Rc2lTZVqHqAYb7vCT2SC/Dvxc13Qj8NbuNutBsS/GG/jzVOa9eGKaXkQxaUsNUU3jJ5lBh/tipvVI3Ieh9YXTQmcluQ/YMbMK/Ey3HqCq6ql91TZp3Vj7RmBmeOY24BeAC6vqv/dY2qIGe0HVi0Wz7It5LgauABbedv9s4FnAv554Rf15Ut8FTJHNrLEhqsGeua/FI/FqsS9mOS1075IcAXyrBhYca3GIarBn7qzBI/Eqsi9mOS20k2QLcBFwN/A64H3AEcBDkpxXVVf0Wd+EPTHJ9d1ygJ/p1qd2iGrI4f6jqnoQoKp+kOT/DDTMwL6Ya9e4Nw0NdFroWxkNzR0KXAWcXlVXJ3ki8AFGw1dDseaGqIY8LOPFoo59MctpobOSXFdVT++Wb6qqJ83Zdu00zhCZpGkfohrymfuaOxKvIvuiU1WfT/JMRtNCX9I138joZpWhTQt9cM7y3y3YNpWBtlrW4hDVYM/cx5n2I/Ek2Rd7GuC00AeA7zP6BPdwRi8Mp1s/uKoO6qu2SUuyndkhqm0sGKKaxk8xg7pANFeSLUk+k+QjSU7o7sa7AbgryWl91zdJ9sWsJOuSnJvkt5I8uWs7o3vmzFt7Lm+iqmpdVT2qqg6pqvXd8sz6YIK9s76qPllVHwa+UVVXA1TVV3qua1FDHpbxYtEs+2LWJcxOC31LksFOC9U8a26IarDDMl4smmVfzPJ5KhpnLQ5RDfnMfc0diVeRfTHLaaHaQ1Wt67uGfTXkM/c1dyReLfbFLKeFqhWDPXNfi0fi1WJfzOO0UDVhsGfu0nI5LVRr0WCnQkrjOC1UrfDMXZpjLd6sIo3jmbs035q7WUUax3CX5nNaqJrgsIw0h9NC1QrDXZIa5LCMJDXIcJekBhnuktQgw12SGmS4S1KD/j+BeR46gkw/fQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Patients distribution\n",
    "pd.Series(a).value_counts(sort=False).plot(kind='bar')\n",
    "plt.show()\n",
    "\n",
    "\n",
    "#Option to upsample Normal and Her2 classes, as under-represented in the set\n",
    "#su = SMOTE(random_state=42)\n",
    "#X_su, y_su = su.fit_resample(df_gene_full__FINAL_transpose, label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "7e344422",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\users\\laura\\appdata\\local\\programs\\python\\python36\\lib\\site-packages\\sklearn\\svm\\_base.py:986: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.\n",
      "  \"the number of iterations.\", ConvergenceWarning)\n"
     ]
    }
   ],
   "source": [
    "# Feature selection to delete noice, and reduce size of the set\n",
    "clf =LinearSVC(C=0.01, penalty=\"l1\", dual=False)\n",
    "clf = clf.fit(df_gene_full__FINAL_transpose, a)\n",
    "model = SelectFromModel(clf, prefit=True)\n",
    "X_new = model.transform(df_gene_full__FINAL_transpose)\n",
    "X_new = pd.DataFrame(X_new)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "10f5485b",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Machine learning model , via logistic regression, solver lbfgs to save memorhy, and multinomial class and give overall better results than other solvers. Penalty l2 according to solver\n",
    "#Set are splitted with StratifiedKFold to avoid overfitting while keeping balanced set\n",
    "#Possibility to also train the model with cross_validate, and pileine including Features selection\n",
    "acc_score = []\n",
    "\n",
    "kf = StratifiedKFold(n_splits = 5)\n",
    "logreg = LogisticRegression(max_iter = 800,multi_class='multinomial', solver='lbfgs')\n",
    "\n",
    "for train_index, test_index in kf.split(X_new,a):\n",
    "     X_train, X_test = X_new.iloc[train_index], X_new.iloc[test_index]\n",
    "     y_train, y_test = a[train_index], a[test_index]\n",
    "     logreg.fit(X_train, y_train)\n",
    "     y_pred = logreg.predict(X_test)\n",
    "     accuracy_score = logreg.score(X_test, y_test)\n",
    "     acc_score.append(accuracy_score)\n",
    "     coeff_log = logreg.coef_\n",
    "     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "ac5c9b69",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy of logistic regression classifier on test set: 0.96\n",
      "[[33  0  0  1  0]\n",
      " [ 0 14  0  2  0]\n",
      " [ 0  0 97  1  1]\n",
      " [ 0  0  1 38  0]\n",
      " [ 0  1  0  0  6]]\n"
     ]
    }
   ],
   "source": [
    "#Model score\n",
    "#Overall accuracy above 85%\n",
    "print('Accuracy of logistic regression classifier on test set: {:.2f}'.format(mean(acc_score)))\n",
    "print(confusion_matrix(y_test, logreg.predict(X_test)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7e36eb42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.08558305180505409\n",
      "0.10694279268475453\n",
      "0.13457794099457118\n",
      "0.11967022318294263\n",
      "0.11041303111014934\n"
     ]
    }
   ],
   "source": [
    "# Predicive gene for each class in list_coeff_sup\n",
    "list_coeff_sup = []\n",
    "\n",
    "for inde_coeef in coeff_log :\n",
    "    list_new = []\n",
    "    for indice_weight in range(0, len(inde_coeef)):\n",
    "        if inde_coeef[indice_weight] > 0.030 :\n",
    "            \n",
    "            list_new.append(indice_weight)\n",
    "    list_coeff_sup.append(list_new)\n",
    "print(list_coeff_sup)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "cf298e08",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Logistic model was chosen from SVM , XGBoost and decision,as more performant for these data. \n",
    "# Deep learning models were discarded as predictive genes more complicated to isolate\n",
    "\n",
    "#Model were test without and with various features selection algorithms were test, but best resuts with LinearSVC\n",
    "#Recursive feature selection can also be tested, but computionally  intensive\n",
    "#SMOTE can also be included to upsample and increase patients with normal and Her2 cancer\n",
    "#Traning should then be done with X_su, y_su\n",
    "#From the model it is clear that the correlation between the subtype and genetic information is quite high\n",
    "#However, for genes with same name but different letter A, B or C, only the A occurences were kept. Other option consists on copying subtype instances for each genes' iteration.\n",
    "\n",
    "\n",
    "\n",
    "# Predictive genes were isolated by isolating coefficient above 0.030\n",
    "#Better gene selection can be done by analysing genes with highest associated weights, and deviation inbetween the weights  for each class\n",
    "\n",
    "\n",
    "#The model could be useful in real life with a larger data set, with more instance of Normal and Her2 subtype for example, covering more variants.\n",
    "#Model also need some adjusting in selecting predictive genes as stated above, to understand how many genes are responsible for the classification, and comparison in between classes. As some might have very similar predictive signature\n",
    "\n",
    "# Futher investigation need to be concluded, comparing predictive genes for each class\n",
    "#and analysing associated protein and their roles.\n",
    "\n",
    "#With more time or resources, I would like to employ Graph convolutional neural networks\n",
    "#Gene does not work individually but as an ensemble, GCNN can learn from already known biological interaction and recognize their patterns. Spatial gene expression can then be retrieved from biopsie, or tissue sample.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b2556f8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
